src/lib.rs

```diff
@@ -1080,6 +1091,14 @@ pub fn sample<T, I, R>(rng: &mut R, iterable: I, amount: usize) -> Vec<T>
         .unwrap_or_else(|e| e)
 }
 
+#[cfg(feature="std")]
+fn log_cause(mut e: &::std::error::Error) {
```

Comment:
Do we know enough about the error types to be able to format them using a `Debug` impl including the chain of causes instead of multiple separate logging statements?

Reply:
`Debug` is for dumping, `Display` for pretty-printing, as I understand. This code attempts to do the latter, which requires multiple lines.

And yes, mostly we're using the `Error` type from the `rand::error` module.

Reply:
Ah gotcha. It might still be worth combining the multiple log statements into 1, otherwise you might end up with your chain of causes interleaved with other log messages.

Reply:
I'm struggling to find a good middle-ground. Compare:
```
ERROR 2018-01-30T12:57:19Z: rand: JitterRng failed: RNG error [permanent failure or unavailable]: timer jitter failed basic quality tests
 WARN 2018-01-30T12:57:19Z: rand: cause: no timer available
```
and
```
ERROR 2018-01-30T12:59:17Z: rand: JitterRng failed: RNG error [permanent failure or unavailable]: timer jitter failed basic quality tests; full error: Error { kind: Unavailable, msg: "timer jitter failed basic quality tests", cause: Some(NoTimer) }
```

Reply:
Personally, I prefer the second one even though it doesn't look as nice in that example because the log message holds all the details. In the same samples with other log messages interleaved the second is easier to follow:

```
ERROR 2018-01-30T12:57:19Z: rand: JitterRng failed: RNG error [permanent failure or unavailable]: timer jitter failed basic quality tests
 INFO 2018-01-30T12:57:19Z: some_crate::some_mod: request to `/`
DEBUG 2018-01-30T12:57:19Z: some_crate::some_mod::some_other_mod: getting a connection to the database
DEBUG 2018-01-30T12:57:19Z: some_crate::some_mod::some_other_mod: got a connection to the database
TRACE 2018-01-30T12:57:19Z: rand: Seeding new RNG
 WARN 2018-01-30T12:57:19Z: rand: cause: no timer available
 INFO 2018-01-30T12:57:19Z: some_crate::some_mod: processing request `/` with handler `Root`
```

```
ERROR 2018-01-30T12:57:19Z: rand: JitterRng failed: RNG error [permanent failure or unavailable]: timer jitter failed basic quality tests; full error: Error { kind: Unavailable, msg: "timer jitter failed basic quality tests", cause: Some(NoTimer) }
 INFO 2018-01-30T12:57:19Z: some_crate::some_mod: request to `/`
DEBUG 2018-01-30T12:57:19Z: some_crate::some_mod::some_other_mod: getting a connection to the database
DEBUG 2018-01-30T12:57:19Z: some_crate::some_mod::some_other_mod: got a connection to the database
TRACE 2018-01-30T12:57:19Z: rand: Seeding new RNG
 INFO 2018-01-30T12:57:19Z: some_crate::some_mod: processing request `/` with handler `Root`
```